Don't suggest adding return type for closures with default return type

wafarm opened the pull request:
<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r​? <reviewer name>
-->
<!-- homu-ignore:end -->

Follow up of #129223

r? @compiler-errors

wafarm committed:
Don't suggest adding return type for closures with default return type

compiler-errors commented on compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs:
```diff
@@ -832,7 +832,11 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 return true;
             }
-            &hir::FnRetTy::DefaultReturn(span) if expected.is_unit() => {
+            // Don't suggest adding return types for closures with default return.
```
There's some redundancy between this check and the `can_suggest` boolean above. Can you find out if we can just reuse `can_suggest` in this position, rather than making the check from within the function?

wafarm replied:
I did some investigations and I found that 

https://github.com/rust-lang/rust/blob/bf662eb80838008acabc307dd64d84935ce3a20d/compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs#L940

If I set `can_suggest` to `false` for closures here, then for `tests/ui/closures/add_semicolon_non_block_closure.rs`

https://github.com/rust-lang/rust/blob/bf662eb80838008acabc307dd64d84935ce3a20d/compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs#L829-L835

code here will go to the upper branch. I think we still need some way to distinguish `main` and closures or we will get something like this:

```
 --> testfile.rs:8:12
  |
8 |     foo(|| bar())
  |           -^^^^^ expected `()`, found `i32`
  |           |
  |           expected `()` because of default return type
  |
```

I don't expect the last line to be printed.

I've thought of some possible solutions:
- The above result is acceptable?
- Keep my original solution, which indeed adds to complexity
- Add `is_main` to the return values of `get_fn_decl`, which seems (I think) a bit overkill and may impact performance if it's getting called a lot
- Refactor some of the code to see if we can find a better solution

compiler-errors reviewed (changes requested):
Would appreciate testing out the suggestion I had.

wafarm commented:
I removed `can_suggest` from `get_fn_decl` as the only usage of it is to be passed to `suggest_missing_return_type`. Instead, I do the check in a new function `can_add_return_type`.

This also fixed not suggesting adding return type in non-trait `impl` blocks.

@rustbot ready

compiler-errors commented:
@bors r+

bors commented:
:pushpin: Commit 736ab66c62ad23b880299e1f997ff1e4d7a9ceef has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 736ab66c62ad23b880299e1f997ff1e4d7a9ceef -->
<!-- homu: {"type":"Approved","sha":"736ab66c62ad23b880299e1f997ff1e4d7a9ceef","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
